synapse/http/client.py

```diff
@@ -386,26 +386,23 @@ def __init__(self, hs):
 
     def endpointForURI(self, uri):
         logger.info("Getting endpoint for %s", uri.toBytes())
+
         if uri.scheme == "http":
-            return SpiderEndpoint(
-                reactor, uri.host, uri.port, self.blacklist, self.whitelist,
-                endpoint=TCP4ClientEndpoint,
-                endpoint_kw_args={
-                    'timeout': 15
-                },
-            )
+            endpoint_factory = HostnameEndpoint
         elif uri.scheme == "https":
-            tlsPolicy = self.policyForHTTPS.creatorForNetloc(uri.host, uri.port)
-            return SpiderEndpoint(
-                reactor, uri.host, uri.port, self.blacklist, self.whitelist,
-                endpoint=SSL4ClientEndpoint,
-                endpoint_kw_args={
-                    'sslContextFactory': tlsPolicy,
-                    'timeout': 15
-                },
-            )
+            tlsCreator = self.policyForHTTPS.creatorForNetloc(uri.host, uri.port)
+
+            def endpoint_factory(reactor, host, port, **kw):
+                return wrapClientTLS(
+                    tlsCreator,
+                    HostnameEndpoint(reactor, host, port, **kw))
         else:
             logger.warn("Can't get endpoint for unrecognised scheme %s", uri.scheme)
+            return None
+        return SpiderEndpoint(
+            reactor, uri.host, uri.port, self.blacklist, self.whitelist,
+            endpoint=endpoint_factory, endpoint_kw_args=dict(timeout=15),
```

Comment:
The rationale is as follows:

1. This dictionary is going to be used as `**kw`, meaning it needs to be native `str` mapped to some object.
2. To facilitate python 3 portability, all modules should slowly be getting as many `__future__` imports as possible, to make the py2 environment and the py3 environment consistent.  One such import is `from __future__ import unicode_literals`.
3. If you rely on `""` to produce a native `str`, this assumption will be broken with `unicode_literals` imported.
4. However, `dict(**kw)` will produce `kw` unaltered; meaning, it will be a native `str` even if a quoted string would produce a `unicode`.

So this is generally a habit I believe one should get into for py3 hygiene.